Step1 - 인수 테스트 기반 리팩터링

sangjaeoh opened the pull request:
안녕하세요 리뷰어님~😃

인수 테스트 기반 리팩터링 PR요청드립니다. 🙇🏻‍♂️

감사합니다 :)

sangjaeoh committed:
docs: 요구사항 작성

sangjaeoh committed:
refactor: LineService getStations

sangjaeoh committed:
refactor: LineResponse

sangjaeoh committed:
refactor: LineService updateLine

sangjaeoh committed:
refactor: LineService addLineStation

sangjaeoh committed:
refactor: LineService removeLineStation

sangjaeoh committed:
refactor: LineService @Transactional

sangjaeoh committed:
refactor: Sections getStations

sangjaeoh committed:
refactor: Line 정적 팩토리 메서드 사용

sangjaeoh committed:
refactor: Section 정적 팩토리 메서드 사용

sangjaeoh committed:
refactor: Sections 정적 팩토리 메서드 사용

sangjaeoh committed:
refactor: LineService saveLine

sangjaeoh committed:
refactor: LineService addLineStation

sangjaeoh committed:
test: 인수 테스트 통합

sangjaeoh committed:
refactor: 메소드명 수정

sangjaeoh committed:
feat: GlobalExceptionHandler 추가

- RuntimeException을 IllegalArgumentException으로 변경
- error code 500 -> 400 으로 반환

sangjaeoh committed:
refactor: StationService 중복 제거

- findById제거, findStationById 사용

wheejuni commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -1,182 +1,75 @@
-
 @Service
-@Transactional
+@Transactional(readOnly = true)
```
딱 봤을때 `readOnly = false` 여야 할 로직이 더 많아 보이긴 하는데요, 여기에 이 옵션을 부여하신 이유가 어떤 것일까요?

sangjaeoh replied:
조회만 필요한 메서드에서 혹시 모를 데이터 변경하는 일을 방지하기 위해 작성하고,
데이터 변경이 필요한 곳에서 직접 `@Transactional`을 다시 달아주고 있습니다!

wheejuni replied:
혹시 모를 데이터 변경하는 일이 발생했다면, 객체에는 여전히 값이 묻어있는 상태일 것 같은데 그렇게 안전장치로써 실효성이 완전히 커 보이지는 않기는 합니다.

wheejuni commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -1,182 +1,75 @@
 
     public Line findLineById(Long id) {
-        return lineRepository.findById(id).orElseThrow(RuntimeException::new);
+        return lineRepository.findById(id).orElseThrow(IllegalArgumentException::new);
```
`IllegalArgumentException` 을 대신할 다른 커스텀 예외를 설계해 주시면 좋겠군요.

wheejuni commented on src/main/java/nextstep/subway/common/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,16 @@
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler({DataIntegrityViolationException.class, IllegalArgumentException.class})
```
`IllegalArgumentException` 은 Java가 기본 제공하는 예외로 프로젝트 로직에서만 한정되게 발생한다고 확신하긴 조금 어려울 것 같기는 합니다. 그래서 아래 리뷰에서 이것을 대신할, 이 프로젝트에서만 사용할 예외를 작성 부탁 드렸습니다.

검토해 보시고 반영해주세요~

wheejuni commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -1,43 +1,62 @@
+        return new Line();
+    }
+
+    public static Line of(String name, String color) {
+        return new Line(name, color);
+    }
+
+    public static Line of(String name, String color, Station upStation, Station downStation, int distance) {
+        return new Line(name, color, upStation, downStation, distance);
```
builder 패턴을 적용해 보시는 것이 조금 더 효율적일 것 같습니다.

sangjaeoh replied:
`builder` 패턴으로 사용하도록 변경하였습니다!

다만 테스트 코드에서 간결하게 사용하기에는 팩토리 메서드가 `builder` 보다 좋다고 생각하여
기존의 팩토리 메서드를 삭제하지는 않았습니다.

wheejuni replied:
@sangjaeoh 테스트 코드에서 간결하게 사용하는 것이 설계에 침투하는 현상을 어떻게 보시나요?
테스트 코드는 테스트 역할에만 충실해야 하고, 로직을 더 깔끔하게 하는 것에 역할을 하면 안됩니다.
만약 말씀처럼 테스트 코드에서 간결하게 사용해야 하는 것이 이슈라면 오히려 테스트 코드 쪽에서 code smell이 난다고 볼 수도 있겠습니다.
리팩토링 방향을 결정하실 때에 앞으로는 테스트 코드의 간결함은 빼 주시면 좋을 듯 합니다.

wheejuni commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -1,43 +1,62 @@
         this.color = line.getColor();
     }
 
+    public void update(String name, String color) {
```
너무 모호한 메소드명이며, 두 인자가 모두 `String` 타입이라 굉장히 위험한 설계인 것 같습니다. 두 인자가 자리가 바뀌어도 타입 체크가 되지 않습니다.

wheejuni commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,165 @@
+
+    private void updateSection(Section section) {
+        sections.stream()
+                .forEach(it -> it.update(section.getUpStation(), section.getDownStation(), section.getDistance()));
```
`Section` 의 `update()` 메소드가 설계가 잘못된 것 같습니다.
게터로 꺼내지 않고 그냥 `Section` 타입 객체만 받아도 로직이 수행되어야 할 것 같군요.

wheejuni commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,165 @@
+
+        Station currentStation = firstStation;
+        while (currentStation != null) {
+            Optional<Station> nextStation = findNextStation(currentStation);
+            if (!nextStation.isPresent()) {
+                break;
+            }
+            currentStation = nextStation.get();
+            stations.add(nextStation.get());
```
`Optional` 의 `ifPresent` 로 묶는 것은 어려우려나요?

sangjaeoh replied:
적용 완료했습니다!

리팩터링을 하면서 비슷한 모양의 `findFirstStation` 메서드를 발견하여
이것도 `ifPregent`로 변경할 수 있나 고민하였지만
`ifPresent`의 `Consumer`로는 `upStation`을 새로운 `Station`으로 할당할 수 없어서
`isPresent`를 그대로 사용하였습니다!

wheejuni reviewed (changes requested):
수고 많으셨습니다 💯 
몇가지 의견 드리겠습니다.

sangjaeoh committed:
refactor: Line update

- name, color를 받아 업데이트 하는 메소드 삭제

sangjaeoh committed:
refactor: Section update

- sections의 updateSection 함수가 section의 getter를 쓰지 않도록 변경

sangjaeoh committed:
refactor: Sections getSortedStations

- isPresent에서 ifPresent사용으로 변경

findFirstStation 메서드의 경우 isPresent를 그대로 사용하도록 유지

sangjaeoh committed:
refactor: Line Builder

- Builder 추가 및 Line.of -> Builder 이용하도록 변경
- 테스트 코드에서 간결하게 사용하기가 Builder보다 좋다고 판단하여 기존 정적 팩토리 메서드는 삭제 X
- Line.empty도 삭제 X

sangjaeoh committed:
refactor: Section Builder

- Builder 추가 및 Section.of -> Builder 이용하도록 변경
- 테스트 코드에서 간결하게 사용하기가 Builder보다 좋다고 판단하여 기존 정적 팩토리 메서드는 삭제 X

sangjaeoh committed:
refactor: 프로젝트 전용 예외로 변경

wheejuni reviewed (approved):
👍 수고하셨습니다. 코멘트 남기신 부분에 제 생각도 좀 남겨보았는데요, 참고만 부탁드립니다~
